### PR TITLE
Added dismissible and required features to announcements

### PR DESCRIPTION
**Pull request by abujeda**
This is a prototype PR and will require more work to complete the implementation.

Adding dismissible and required features to announcements.
 - **Dismissible Announcements:** Will add a button to the announcement. When executed, the system will store that the announcement was completed in the user settings.

 - **Required Announcements:** The same as dismissible, but will block all request to the application and only display an empty homepage with all the announcements until all required announcements have been executed.

Possbile implementation for #3634

### CODE REVIEW

**Comment by abujeda** [comment]
Some sample announcements:
```yaml
msg: HELLO
required: true
```

```yaml
msg: HELLO
type: warning
dismissible: true
button_text: "Accept"
```

```yaml
id: policy_556234
msg: Policy Text
required: true
button_text: "Accept"
```

**Comment by johrstrom** [comment]
Hey sorry for the delay in getting around to this. Honestly, I kinda feel like all announcements should be dismiss-able. 

I haven't tested this out but will do so shortly. Looking it over though, it seems feature complete. Not sure what you mean by prototype or if that was just a qualifier you had in case I'd requested more or less.

In any case - it looks pretty good on the surface, I just need a sec to play around with it.

**Comment by abujeda** [comment]
Thanks Jeff, great that you like the new features for Announcements

I need to add tests and review the implementation in case I missed an edge case.
It should not change the general approach. I will work on that to complete the PR

File: apps/dashboard/app/models/announcement.rb [hunk 1]
```diff
@@ -31,10 +32,36 @@ def msg
     msg.blank? ? nil : msg
   end
 
+  def button_text
+    opts.fetch(:button_text, I18n.t('dashboard.announcements_button_text')).to_s
```
Comment (johrstrom) [inline apps/dashboard/app/models/announcement.rb:36]: I feel like this should be 2 things, one for `dismissable`, and another for `required`.

Required should be something like `Accept` and `dismissable` maybe, `Don't show again`.

File: apps/dashboard/app/models/announcement.rb [hunk 2]
```diff
@@ -31,10 +32,36 @@ def msg
...
+  end
+
+  def required?
+    opts.fetch(:required, false)
+  end
+
+  def id
+    default_id = @path.basename.to_s if @path
+    # default_id = Digest::SHA1.hexdigest(@path.basename.to_s) if @path
+    opts.fetch(:id, default_id)
```
Comment (johrstrom) [inline apps/dashboard/app/models/announcement.rb:64]: OSC reuses names all the time. We publish the `ondemand_update.yml` all the time and it has a time check within it to show the alert based on the current time.

In any case - I feel like the id should be a hexdigest of the content of the file so folks can continue to reuse the same filename and not have ot say publish _1.yml and _2.yml and so on forever.

Circling back to my feeling that all announcements should be dismissable - by making the id the content of the file we can avoid collisions and having to increment or put dates in filenames.,

File: apps/dashboard/app/models/announcement.rb [hunk 3]
```diff
@@ -31,10 +32,36 @@ def msg
...
+    true
+  end
+
+  def completed?
+    dismissible? && user_settings.dig(:announcements, id.to_sym)
+  end
+
+  def dismissible?
+    required? || opts.fetch(:dismissible, false)
+  end
```
Comment (johrstrom) [inline apps/dashboard/app/models/announcement.rb:55]: I'm wondering if all announcements should be dismissable. I'm thinking yes as they take up a bit of real estate and can be on all pages for days. 

I guess as a best of both worlds - we can just default here to `true` and folks can set `false` if they wish to keep it on the page.

File: apps/dashboard/app/models/announcement.rb (hunk 1 above)
Comment (johrstrom) [reply apps/dashboard/app/models/announcement.rb:36]: 🤦‍♂️ I see now that this is overridable. Even so, maybe we can still supply two defaults depending, just for convenience so they don't have to always override the button text.

File: apps/dashboard/app/models/announcement.rb (hunk 3 above)
Comment (abujeda) [reply apps/dashboard/app/models/announcement.rb]: Fixed

File: apps/dashboard/app/models/announcement.rb (hunk 2 above)
Comment (abujeda) [reply apps/dashboard/app/models/announcement.rb]: I have updated to use the `hexdigest` of the message.

File: apps/dashboard/app/models/announcement.rb (hunk 1 above)
Comment (abujeda) [reply apps/dashboard/app/models/announcement.rb]: I added 2 default button texts, one for required and one for dismissible

**Comment by abujeda** [comment]
I have completed the refactoring and added some tests.

File: apps/dashboard/config/routes.rb [hunk 4]
```diff
@@ -102,7 +102,7 @@
 
   get '/system-status', to: 'system_status#index', as: 'system_status' if Configuration.can_access_system_status?
 
-  post 'settings', :to => 'settings#update'
+  post 'settings/:action', to: 'settings#:action', as: 'settings'
```
Comment (johrstrom) [inline apps/dashboard/config/routes.rb:105]: I can't quite discern why we need this change. Is it because of the id change I requested?

Comment (abujeda) [reply apps/dashboard/config/routes.rb:105]: This change is not related to your suggestions/changes.
This is an improvement I wanted to make to have specific methods in the controller to manage the updates to different types of settings.

The idea being that updating a profile will be managed by the `profile` method. Updating a announcement will be managed by the `announcement` method. This way we can apply logic specific to the type without having to add a lot of conditional logic to the previous single `update` method.

I can revert back to the previous way with the single method update a leave this for a later date.

Comment (abujeda) [reply apps/dashboard/config/routes.rb]: I have reverted the changes. The PR ready for a final review

File: apps/dashboard/app/controllers/settings_controller.rb [hunk 5]
```diff
@@ -1,17 +1,45 @@
...
+        format.json { head :bad_request }
+      end
 
-    logger.info "settings: updated user settings to: #{new_settings}"
+      return
+    end
+
+    # Only ALLOWED_SETTINGS methods will be called
+    setting, value = new_settings.first
+    send("update_#{setting}", value)
```
Comment (johrstrom) [inline apps/dashboard/app/controllers/settings_controller.rb:22]: Yea I feel like this is a bit like `working by coincidence`. This is the first PR to update the settings controller, so we're just at that spot, but I feel like it should be easier. 

That is, I don't think we need different methods for different things we're updating. Or at least I hope not just because that'll explode. 

Can't we follow what we had before? I guess I'm really asking if we need all these special methods or if we can't just merge the 2 hashes and save?

* filter the parameters coming into the controller
* read the current settings
* merge
* save

Comment (abujeda) [reply apps/dashboard/app/controllers/settings_controller.rb:22]: I think we are better defining the contract of settings as we add more. The new contract means that you can only update one setting at the time using the `SettingsController`. And the specific methods is to be able to add more logic into each setting into the backend. Otherwise it will have to move somewhere else.

My first implementation was as you suggested, but I had to allow to submit a hash as a value to support multiple announcements. This is the pseudocode for the that solution:

```ruby
SettingsController
...
  ALLOWED_SETTINGS = [:profile, { announcements: {} }].freeze
...
  def read_settings(params)
    {}.tap do |settings|
      params&.each do |key, value|
        value = value.to_h if value.is_a?(ActionController::Parameters)
        settings[key] = value
      end
    end
  end
```

```erb
# The button to summit the setting
<%=
    button_to(
      settings_path("settings[announcements][#{announcement.id}]" => '2024-07-18 12:00:00'),
      method: :post,
      class: %w[btn btn-warning me-md-2],
    ) do
      'Accept'
    end
  %>
```

Comment (abujeda) [reply apps/dashboard/app/controllers/settings_controller.rb:22]: Actually, now that I look at the code, this could be simplified to:

```ruby
  def read_settings(params)
    params.to_h
  end
```

My only concern is now allowing a hash as a parameter. Do you see this as an issue?

Comment (abujeda) [reply apps/dashboard/app/controllers/settings_controller.rb]: I have updated the SettingsController as per your suggestion. Looking at it now, it is not as complicated as my first implementation. Let me know what you think

Comment (johrstrom) [reply apps/dashboard/app/controllers/settings_controller.rb]: Yea I get all that, that it'll get complicated as we go. IDK about passing a hash into from a `form`, so I guess we'll have to work that out. If it's simple now and works, then I'm happy and we can sort of continue to refine it.

That said - at some point we're likely to make a page so that folks can edit their settings, so it's likely one giant `form` with all possible configurations.  So this is kinda why I'm pushing for simplicity, so that we can provide an `edit` page someday.

**Review by johrstrom — APPROVED** [review]
Thanks!
